### PR TITLE
Alerting: Make Loki & Prometheus instant vector by default

### PR DESCRIPTION
**Pull request by gillesdemey**
**What is this feature?**

This feature makes the following changes to Alerting:

**Non-visually**

* Adds support for `getDefaultQuery()` for datasources
* Makes Prometheus and Loki queries use `instant` responses by default in Alerting

**Visually**
* Removes the `table` and `stat` visualisations as these were broken for dataframes with lots of labels
* Automatically shows the most useful visualization based on the dataframe type (timeseries vs `numeric-multi`)
* Uses the `<GraphContainer />` from the explore view since it has a few features that are useful in Alerting
  * Slicing the set of data frames / series to a reasonable default
  * Toggling series works out-of-the-box
  * Bar / Points vizualizations
  * Loading states
* Uses the existing `<ExpressionResult />` component to render `numeric-multi` dataframes
* Updates the `<ExpressionResult />` to syntax color labels
* Adds threshold visualisations to `<GrafanaRuleQueryViewer />`

And makes the following changes to other parts of Grafana that aren't Alerting:

* Updates the `<GraphContainer />` to support thresholds
* Adds a `onDataSourceLoaded()` hook to `<QueryEditorRow />`

image

**Why do we need this feature?**

This would reduce to TCO for Loki and Mimir by no longer querying a range of data for each alert rule evaluation.

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

~Setting this to _draft_ for now until I can figure out how to write some tests for some of the most important bits here.~

### CODE REVIEW

**Comment by gabor** [comment]
this may not be relevant, but i saw this part:
> Automatically shows the most useful visualization based on the dataframe type (timeseries vs numeric-multi)

i think right now, to geot `numeric-multi` some feature-flags have to be enabled for Loki and Prometheus. is that ok?

**Comment by gillesdemey** [comment]
I'll double check with the alerting BE team if the feature flag is relevant to us, the `numeric-multi` dataframes are returned by the `/api/v1/eval` endpoint and not from the Loki / Prom data sources directly.

File: public/app/plugins/datasource/loki/datasource.ts [hunk 1]
```diff
@@ -824,6 +824,14 @@ export class LokiDatasource
   getQueryHints(query: LokiQuery, result: DataFrame[]): QueryHint[] {
     return getQueryHints(query.expr, result);
   }
+
+  getDefaultQuery(app: CoreApp): Partial<LokiQuery> {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/datasource.ts:828]: Could we do `LokiQuery` instead of `Partial<LokiQuery>` and add default values for all required fields which is I think only `expr`. And also create a defaultQuery for app that is not `UnifiedAlerting`?

**Review by ivanahuckova — COMMENTED** [review]
I reviewed `loki` part of change and added feedback. Let me know what do you think!

File: public/app/plugins/datasource/loki/datasource.ts (hunk 1 above)
Comment (gillesdemey) [reply public/app/plugins/datasource/loki/datasource.ts]: Yep – I think that's sensible 👍  Will apply full defaults for both Loki and Prometheus for consistency.

File: public/app/features/alerting/unified/components/expressions/Expression.tsx [hunk 2]
```diff
@@ -415,12 +440,21 @@ const getStyles = (theme: GrafanaTheme2) => ({
         background-color: ${theme.colors.background.canvas};
       }
     `,
+    labelKey: css`
+      color: ${theme.isDark ? '#73bf69' : '#56a64b'};
```
[the remark this answers is not in the document]
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/expressions/Expression.tsx:444]: I copied them from here https://github.com/grafana/grafana/blob/main/public/app/features/explore/PrometheusListView/RawListItemAttributes.tsx#L10-L12 so I think it's fine to make a few exceptions :D

File: public/app/features/alerting/unified/components/expressions/Expression.tsx [hunk 3]
```diff
@@ -302,16 +302,37 @@ const FrameRow: FC<FrameProps> = ({ frame, index, isAlertCondition }) => {
...
 
+  const title = `${hasLabels ? '' : name}${hasLabels ? `{${formatLabels(labelsRecord)}}` : ''}`;
+
   return (
     <div className={styles.expression.resultsRow}>
       <Stack direction="row" gap={1} alignItems="center">
-        <span className={cx(styles.mutedText, styles.expression.resultLabel)} title={name}>
-          {name}
-        </span>
+        <div className={cx(styles.expression.resultLabel)} title={title}>
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/expressions/Expression.tsx:317]: why do you use `cx` in here instead of directly `styles.expression.resultLabel`?

File: public/app/features/alerting/unified/components/rule-editor/QueryWrapper.tsx [hunk 4]
```diff
@@ -117,67 +115,65 @@ export const QueryWrapper = ({
...
-        }
-        hideDisableQuery={true}
-      />
-    </div>
+    <Stack direction="column" gap={0.5}>
+      <div className={styles.wrapper}>
+        <QueryEditorRow<DataQuery>
+          alerting
+          dataSource={dsSettings}
+          onDataSourceLoaded={setDsInstance}
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/rule-editor/QueryWrapper.tsx:148]: 🙌

File: public/app/features/alerting/unified/components/rule-editor/util.ts [hunk 5]
```diff
@@ -261,6 +260,9 @@ export function getThresholdsForQueries(queries: AlertQuery[]) {
     // now also sort the threshold values, if we don't then they will look weird in the time series panel
     // TODO this doesn't work for negative values for now, those need to be sorted inverse
     thresholds[refId].config.steps.sort((a, b) => a.value - b.value);
+
+    // also make sure we remove any "undefined" values from our steps in case the threshold config is incomplete
+    thresholds[refId].config.steps = thresholds[refId].config.steps.filter((step) => step.value !== undefined);
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/rule-editor/util.ts:265]: is this change needed for this PR changes? or you added as an improvement?

**Review by soniaAguilarPeiron — APPROVED** [review]
LGTM!🚀
I added some non-blocking comments

File: public/app/features/alerting/unified/components/rule-editor/util.ts (hunk 5 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/rule-editor/util.ts:265]: Not strictly need, it fixes a small bug I found while testing my changes where an incomplete threshold definition (like `is between 0 and <undefined>` would break the time series visualization and show a blank canvas.

File: public/app/features/alerting/unified/GrafanaRuleQueryViewer.tsx [hunk 6]
```diff
@@ -46,6 +47,8 @@ export function GrafanaRuleQueryViewer({
   const expressions = queries.filter((q) => isExpressionQuery(q.model));
   const styles = useStyles2(getExpressionViewerStyles);
 
+  const thresholds = getThresholdsForQueries(queries);
```
Comment (konrad147) [inline public/app/features/alerting/unified/GrafanaRuleQueryViewer.tsx:50]: I noticed an error is thrown inside `getThresholdsForQueries` when you switch from the classic condition to another expression type. No UI visible side effects though

**Comment by konrad147** [comment]
Great job. Visualizations are handled much better now! 
I have one general concern about the use of instant vectors - Using them, we're not able to visualize the query as a time series chart which might be helpful in many scenarios (esp., how to select a sensible threshold value)
In the future, we could try to use an instant query for alert definition and range just for the preview

File: public/app/features/alerting/unified/components/rule-editor/VizWrapper.tsx [hunk 7]
```diff
@@ -1,136 +1,86 @@
...
-                />
-              </PanelContextProvider>
-            </div>
-          );
-        }}
+      <AutoSizer disableHeight>
+        {({ width }) => (
+          <div style={{ width }}>
+            {isTimeSeriesData ? (
+              <GraphContainer
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rule-editor/VizWrapper.tsx:36]: It looks like `PanelRenderer` has built-in error handling. As we no longer use it we need to handle error on our own

**Comment by gillesdemey** [comment]
Yea valid concern; unfortunately we have to choose between;

Instant vector – better for querying as fewer data is transferred and is faster, but not time series viz
Range – more data transfer, slower but nice time series viz

Unless we figure out how to rewrite queries  I don't see how we could have both :(

**Review by konrad147 — APPROVED** [review]
LGTM! Great job!

File: public/app/features/alerting/unified/GrafanaRuleQueryViewer.tsx (hunk 6 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/GrafanaRuleQueryViewer.tsx:50]: Seems to be in `main` too but I was able to track it down and fixed it in https://github.com/grafana/grafana/pull/66797/commits/e43dd950d656d2e18eb335b611ba6c8944c4450b :)
